Panic when using Visual Studio extension

### What happened?

Language server crashes with a panic when trying to use against https://github.com/upbound/platform-ref-azure. 

### How can we reproduce it?

- Install VS Code up extension `v0.0.4`
- Use up `v0.6.0`
- Clone https://github.com/upbound/platform-ref-azure
- Open up the `platform-ref-azure` directory in visual studio.

Produces a panic for me:

```
panic: invalid package type supplied

goroutine 23 [running]:
github.com/upbound/up/internal/xpls/server.(*Server).Initialize(0x140003658c0, 0x102986878, 0x140001a2008, 0x140001715f0, 0x0, 0x0, 0x0, 0xc00, 0x140001fc400)
	/home/runner/work/up/up/internal/xpls/server/server.go:124 +0x2e8
github.com/upbound/up/internal/xpls/dispatcher.(*Dispatcher).Dispatch(0x1400044b970, 0x102986878, 0x140001a2008, 0x1029919b8, 0x140003658c0, 0x140001715f0, 0x140004fce60)
	/home/runner/work/up/up/internal/xpls/dispatcher/dispatcher.go:80 +0x124
github.com/upbound/up/internal/xpls/handler.(*Handler).Handle(0x14000492be0, 0x102986878, 0x140001a2008, 0x140001715f0, 0x140004fce60)
	/home/runner/work/up/up/internal/xpls/handler/handler.go:69 +0x60
github.com/sourcegraph/jsonrpc2.(*Conn).readMessages(0x140001715f0, 0x102986878, 0x140001a2008)
	/home/runner/work/up/up/vendor/github.com/sourcegraph/jsonrpc2/jsonrpc2.go:553 +0x4dc
created by github.com/sourcegraph/jsonrpc2.NewConn
	/home/runner/work/up/up/vendor/github.com/sourcegraph/jsonrpc2/jsonrpc2.go:334 +0x198
```

--- a/internal/xpkg/dep/manager/utils.go
+++ b/internal/xpkg/dep/manager/utils.go
@@ -16,6 +16,7 @@ package manager
 
 import (
 	metav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	metav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
 	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
 )
 
@@ -35,4 +36,20 @@ func ConvertToV1beta1(in metav1.Dependency) v1beta1.Dependency {
 	}
 
 	return betaD
+}
+
+// ConvertToV1alpha1 converts v1.Dependency types to v1alpha1.Dependency types.
+func ConvertToV1alpha1(in metav1.Dependency) metav1alpha1.Dependency {
+	alphaD := metav1alpha1.Dependency{
+		Version: in.Version,
+	}
+	if in.Provider != nil && in.Configuration == nil {
+		alphaD.Provider = in.Provider
+	}
+
+	if in.Configuration != nil && in.Provider == nil {
+		alphaD.Configuration = in.Configuration
+	}
+
+	return alphaD
 }

--- a/internal/xpkg/snapshot/deprecation_test.go
+++ b/internal/xpkg/snapshot/deprecation_test.go
@@ -1,0 +1,116 @@
+// Copyright 2022 Upbound Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/test"
+	metav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	metav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
+
+	"github.com/upbound/up/internal/xpkg/snapshot/validator"
+)
+
+func TestAPIVersionDeprecation(t *testing.T) {
+	type args struct {
+		o runtime.Object
+	}
+	type want struct {
+		err *validator.Validation
+	}
+
+	cases := map[string]struct {
+		reason string
+		args   args
+		want   want
+	}{
+		"V1alpha1Configuration": {
+			reason: "Should return a deprecation warning.",
+			args: args{
+				o: &metav1alpha1.Configuration{
+					TypeMeta: v1.TypeMeta{
+						Kind:       metav1alpha1.ConfigurationKind,
+						APIVersion: metav1alpha1.SchemeGroupVersion.String(),
+					},
+				},
+			},
+			want: want{
+				err: &validator.Validation{
+					TypeCode: validator.WarningTypeCode,
+					Message:  "meta.pkg.crossplane.io/v1alpha1 is deprecated in favor of meta.pkg.crossplane.io/v1",
+					Name:     "apiVersion",
+				},
+			},
+		},
+		"V1alpha1Provider": {
+			reason: "Should return a deprecation warning.",
+			args: args{
+				o: &metav1alpha1.Provider{
+					TypeMeta: v1.TypeMeta{
+						Kind:       metav1alpha1.ProviderKind,
+						APIVersion: metav1alpha1.SchemeGroupVersion.String(),
+					},
+				},
+			},
+			want: want{
+				err: &validator.Validation{
+					TypeCode: validator.WarningTypeCode,
+					Message:  "meta.pkg.crossplane.io/v1alpha1 is deprecated in favor of meta.pkg.crossplane.io/v1",
+					Name:     "apiVersion",
+				},
+			},
+		},
+		"V1Configuration": {
+			reason: "Should return a deprecation warning.",
+			args: args{
+				o: &metav1.Configuration{
+					TypeMeta: v1.TypeMeta{
+						Kind:       metav1.ConfigurationKind,
+						APIVersion: metav1.SchemeGroupVersion.String(),
+					},
+				},
+			},
+		},
+		"V1Provider": {
+			reason: "Should return a deprecation warning.",
+			args: args{
+				o: &metav1.Provider{
+					TypeMeta: v1.TypeMeta{
+						Kind:       metav1.ProviderKind,
+						APIVersion: metav1.SchemeGroupVersion.String(),
+					},
+				},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+
+			err := validateAPIVersion(tc.args.o)
+
+			if err != nil {
+				if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+					t.Errorf("\n%s\nAPIVersionDeprecation(...): -want error, +got error:\n%s", tc.reason, diff)
+				}
+			}
+		})
+	}
+}

--- a/internal/xpkg/snapshot/meta.go
+++ b/internal/xpkg/snapshot/meta.go
@@ -22,12 +22,14 @@ import (
 	"io/ioutil"
 	"strings"
 
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/kube-openapi/pkg/validation/validate"
 	"sigs.k8s.io/yaml"
 
 	"github.com/Masterminds/semver"
 	"github.com/crossplane/crossplane-runtime/pkg/parser"
-	pkgmetav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	metav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	"github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
 	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
 
 	"github.com/upbound/up/internal/xpkg"
@@ -37,13 +39,15 @@ import (
 )
 
 const (
+	apiVersionField  = "apiVersion"
 	dependsOnPathFmt = "spec.dependsOn[%d].%s"
 	versionField     = "version"
 
-	errFailedConvertToPkg = "unable to convert to package"
-	errPackageDNEFmt      = "Package %s does not exist locally. Please run `up xpkg dep` to fix."
-	errVersionDENFmt      = "Version matching %s does not exist locally. Please run `up xpkg dep` to fix."
-	errWrongPkgTypeFmt    = "Incorrect package type. '%s' does not match type for %s of '%s'"
+	errAPIVersionDeprecatedFmt = "%s is deprecated in favor of %s"
+	errFailedConvertToPkg      = "unable to convert to package"
+	errPackageDNEFmt           = "Package %s does not exist locally. Please run `up xpkg dep` to fix."
+	errVersionDENFmt           = "Version matching %s does not exist locally. Please run `up xpkg dep` to fix."
+	errWrongPkgTypeFmt         = "Incorrect package type. '%s' does not match type for %s of '%s'"
 )
 
 // Validator defines a validator for meta files.
@@ -76,13 +80,16 @@ func DefaultMetaValidators(s *Snapshot) (*Validator, error) {
 // Validate performs validation rules on the given data input per the rules
 // defined for the Validator.
 func (m *Validator) Validate(data interface{}) *validate.Result {
-	pkg, err := m.Marshal(data)
+	pkg, o, err := m.Marshal(data)
 	if err != nil {
 		// TODO(@tnthornton) add debug logging
 		return validator.Nop
 	}
 
 	errs := make([]error, 0)
+
+	// validate the current apiVersion of the meta file
+	errs = append(errs, validateAPIVersion(o))
 
 	for i, d := range pkg.GetDependencies() {
 		cd := manager.ConvertToV1beta1(d)
@@ -97,31 +104,59 @@ func (m *Validator) Validate(data interface{}) *validate.Result {
 }
 
 // Marshal marshals the given data object into a Pkg, errors otherwise.
-func (m *Validator) Marshal(data interface{}) (pkgmetav1.Pkg, error) {
+func (m *Validator) Marshal(data interface{}) (metav1.Pkg, runtime.Object, error) {
 	b, err := yaml.Marshal(data)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// convert data to a package
 	ppkg, err := m.p.Parse(context.Background(), ioutil.NopCloser(bytes.NewReader(b)))
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if len(ppkg.GetMeta()) != 1 {
-		return nil, err
+		return nil, nil, err
 	}
 
-	pkg, ok := xpkg.TryConvertToPkg(ppkg.GetMeta()[0], &pkgmetav1.Provider{}, &pkgmetav1.Configuration{})
+	o := ppkg.GetMeta()[0]
+	pkg, ok := xpkg.TryConvertToPkg(o, &metav1.Provider{}, &metav1.Configuration{})
 	if !ok {
-		return nil, errors.New(errFailedConvertToPkg)
+		return nil, nil, errors.New(errFailedConvertToPkg)
 	}
-	return pkg, nil
+	return pkg, o, nil
 }
 
 type metaValidator interface {
 	validate(int, v1beta1.Dependency) error
+}
+
+// validateAPIVersion tests if the provided object is a deprecated version
+func validateAPIVersion(o runtime.Object) error {
+	switch o.(type) {
+	case *v1alpha1.Configuration:
+		return &validator.Validation{
+			Name: apiVersionField,
+			Message: fmt.Sprintf(
+				errAPIVersionDeprecatedFmt,
+				o.GetObjectKind().GroupVersionKind().GroupVersion(),
+				metav1.ConfigurationGroupVersionKind.GroupVersion(),
+			),
+			TypeCode: validator.WarningTypeCode,
+		}
+	case *v1alpha1.Provider:
+		return &validator.Validation{
+			Name: apiVersionField,
+			Message: fmt.Sprintf(
+				errAPIVersionDeprecatedFmt,
+				o.GetObjectKind().GroupVersionKind().GroupVersion(),
+				metav1.ProviderGroupVersionKind.GroupVersion(),
+			),
+			TypeCode: validator.WarningTypeCode,
+		}
+	}
+	return nil
 }
 
 // TypeValidator validates the dependency type matches the supplied dependency
@@ -144,7 +179,7 @@ func (v *TypeValidator) validate(i int, d v1beta1.Dependency) error {
 		return nil
 	}
 	if got.Type() != d.Type {
-		return &validator.MetaValidation{
+		return &validator.Validation{
 			Name: fmt.Sprintf(dependsOnPathFmt, i, strings.ToLower(string(d.Type))),
 			Message: fmt.Sprintf(errWrongPkgTypeFmt,
 				strings.ToLower(string(d.Type)),
@@ -180,13 +215,13 @@ func (v *VersionValidator) validate(i int, d v1beta1.Dependency) error {
 		return nil // nolint:nilerr
 	}
 	if len(vers) == 0 {
-		return &validator.MetaValidation{
+		return &validator.Validation{
 			Name:    fmt.Sprintf(dependsOnPathFmt, i, strings.ToLower(string(d.Type))),
 			Message: fmt.Sprintf(errPackageDNEFmt, d.Package),
 		}
 	}
 	if !versionMatch(d.Constraints, vers) {
-		return &validator.MetaValidation{
+		return &validator.Validation{
 			Name:    fmt.Sprintf(dependsOnPathFmt, i, versionField),
 			Message: fmt.Sprintf(errVersionDENFmt, d.Constraints),
 		}

--- a/internal/xpkg/snapshot/snapshot.go
+++ b/internal/xpkg/snapshot/snapshot.go
@@ -451,7 +451,7 @@ func validationDiagnostics(res *validate.Result, n ast.Node, gvk schema.GroupVer
 				message: fmt.Sprintf("%s (%s)", et.Error(), gvk),
 				name:    et.Name,
 			}
-		case *validator.MetaValidation:
+		case *validator.Validation:
 			e = &verror{
 				code:    et.Code(),
 				message: et.Error(),
@@ -497,6 +497,18 @@ func validationDiagnostics(res *validate.Result, n ast.Node, gvk schema.GroupVer
 				default:
 					endCh = tok.Position.Column + len(tok.Value) - 1
 				}
+
+				// handle different types of diagnostic notifications
+				var sev protocol.DiagnosticSeverity
+				switch c := e.code; {
+				case c == validator.WarningTypeCode:
+					sev = protocol.SeverityWarning
+				case c == validator.ErrorTypeCode:
+					sev = protocol.SeverityError
+				case c >= 422:
+					sev = protocol.SeverityError
+				}
+
 				// TODO(hasheddan): token position reflects file line
 				// and column by NOT being zero-indexed, but VSCode
 				// interprets ranges with zero-indexing. We should
@@ -513,7 +525,7 @@ func validationDiagnostics(res *validate.Result, n ast.Node, gvk schema.GroupVer
 						},
 					},
 					Message:  e.Error(),
-					Severity: protocol.SeverityError,
+					Severity: sev,
 					Source:   serverName,
 				})
 			}

--- a/internal/xpkg/snapshot/validator.go
+++ b/internal/xpkg/snapshot/validator.go
@@ -31,8 +31,8 @@ import (
 
 	xpextv1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
 	xpextv1beta1 "github.com/crossplane/crossplane/apis/apiextensions/v1beta1"
-	v1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
-	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
+	metav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	metav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
 
 	"github.com/upbound/up/internal/xpkg/snapshot/validator"
 )
@@ -64,20 +64,20 @@ func (s *Snapshot) ValidatorsForObj(o runtime.Object) (map[schema.GroupVersionKi
 		if err := validatorsFromV1XRD(rd, validators); err != nil {
 			return nil, err
 		}
-	case *v1.Configuration:
+	case *metav1.Configuration:
 		if err := s.validatorsForV1Configuration(rd, validators); err != nil {
 			return nil, err
 		}
-	case *v1beta1.Configuration:
-		if err := s.validatorsForV1Beta1Configuration(rd, validators); err != nil {
+	case *metav1alpha1.Configuration:
+		if err := s.validatorsForV1Alpha1Configuration(rd, validators); err != nil {
 			return nil, err
 		}
-	case *v1.Provider:
+	case *metav1.Provider:
 		if err := s.validatorsForV1Provider(rd, validators); err != nil {
 			return nil, err
 		}
-	case *v1beta1.Provider:
-		if err := s.validatorsForV1Beta1Provider(rd, validators); err != nil {
+	case *metav1alpha1.Provider:
+		if err := s.validatorsForV1Alpha1Provider(rd, validators); err != nil {
 			return nil, err
 		}
 	default:
@@ -167,7 +167,7 @@ func validatorsFromV1XRD(x *xpextv1.CompositeResourceDefinition, acc map[schema.
 	return nil
 }
 
-func (s *Snapshot) validatorsForV1Configuration(c *v1.Configuration, acc map[schema.GroupVersionKind]validator.Validator) error {
+func (s *Snapshot) validatorsForV1Configuration(c *metav1.Configuration, acc map[schema.GroupVersionKind]validator.Validator) error {
 	v, err := DefaultMetaValidators(s)
 	if err != nil {
 		return err
@@ -176,7 +176,7 @@ func (s *Snapshot) validatorsForV1Configuration(c *v1.Configuration, acc map[sch
 	return nil
 }
 
-func (s *Snapshot) validatorsForV1Beta1Configuration(c *v1beta1.Configuration, acc map[schema.GroupVersionKind]validator.Validator) error {
+func (s *Snapshot) validatorsForV1Alpha1Configuration(c *metav1alpha1.Configuration, acc map[schema.GroupVersionKind]validator.Validator) error {
 	v, err := DefaultMetaValidators(s)
 	if err != nil {
 		return err
@@ -185,7 +185,7 @@ func (s *Snapshot) validatorsForV1Beta1Configuration(c *v1beta1.Configuration, a
 	return nil
 }
 
-func (s *Snapshot) validatorsForV1Provider(c *v1.Provider, acc map[schema.GroupVersionKind]validator.Validator) error {
+func (s *Snapshot) validatorsForV1Provider(c *metav1.Provider, acc map[schema.GroupVersionKind]validator.Validator) error {
 	v, err := DefaultMetaValidators(s)
 	if err != nil {
 		return err
@@ -194,7 +194,7 @@ func (s *Snapshot) validatorsForV1Provider(c *v1.Provider, acc map[schema.GroupV
 	return nil
 }
 
-func (s *Snapshot) validatorsForV1Beta1Provider(c *v1beta1.Provider, acc map[schema.GroupVersionKind]validator.Validator) error {
+func (s *Snapshot) validatorsForV1Alpha1Provider(c *metav1alpha1.Provider, acc map[schema.GroupVersionKind]validator.Validator) error {
 	v, err := DefaultMetaValidators(s)
 	if err != nil {
 		return err

--- a/internal/xpkg/snapshot/validator/validator.go
+++ b/internal/xpkg/snapshot/validator/validator.go
@@ -18,6 +18,20 @@ import (
 	"k8s.io/kube-openapi/pkg/validation/validate"
 )
 
+// All code responses can be used to differentiate errors for different handling
+// by the consumer
+const (
+	// WarningTypeCode indicates a warning is being returned.
+	WarningTypeCode = 100
+	// ErrorTypeCode indicates an error is being returned.
+	ErrorTypeCode = 500
+
+	// NOTE(@tnthornton) api-server uses error code 422 and 600+ to indicate
+	// validation errors. As long as we're deferring to their logic for
+	// k8s type validations, we need to be sure to not overload those error
+	// codes.
+)
+
 // Nop is used for no-op validator results.
 var Nop = &validate.Result{}
 
@@ -26,19 +40,19 @@ type Validator interface {
 	Validate(data interface{}) *validate.Result
 }
 
-// MetaValidation represents a failure of a meta file condition.
-type MetaValidation struct {
-	code    int32
-	Message string
-	Name    string
+// Validation represents a failure of a file condition.
+type Validation struct {
+	TypeCode int32
+	Message  string
+	Name     string
 }
 
 // Code returns the code corresponding to the MetaValidation.
-func (e *MetaValidation) Code() int32 {
-	return e.code
+func (e *Validation) Code() int32 {
+	return e.TypeCode
 }
 
-func (e *MetaValidation) Error() string {
+func (e *Validation) Error() string {
 	return e.Message
 }
 

--- a/internal/xpkg/workspace/meta/meta.go
+++ b/internal/xpkg/workspace/meta/meta.go
@@ -23,15 +23,17 @@ import (
 	sigsyaml "sigs.k8s.io/yaml"
 
 	v1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	"github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
 	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
 
+	"github.com/upbound/up/internal/xpkg"
 	"github.com/upbound/up/internal/xpkg/dep/manager"
 )
 
 const (
-	errInvalidMetaFile     = "invalid meta type supplied"
-	errMetaContainsDupeDep = "meta file contains duplicate dependency"
-	errNotAPackage         = "invalid package type supplied"
+	errInvalidMetaFile           = "invalid meta type supplied"
+	errMetaContainsDupeDep       = "meta file contains duplicate dependency"
+	errUnsupportedPackageVersion = "unsupported package version supplied"
 )
 
 // Meta provides helpful methods for interacting with a metafile's
@@ -50,9 +52,9 @@ func New(obj runtime.Object) *Meta {
 
 // DependsOn returns a slice of v1beta1.Dependency that this workspace depends on.
 func (m *Meta) DependsOn() ([]v1beta1.Dependency, error) {
-	pkg, ok := m.obj.(v1.Pkg)
+	pkg, ok := xpkg.TryConvertToPkg(m.obj, &v1.Provider{}, &v1.Configuration{})
 	if !ok {
-		return nil, errors.New(errNotAPackage)
+		return nil, errors.New(errUnsupportedPackageVersion)
 	}
 
 	out := make([]v1beta1.Dependency, len(pkg.GetDependencies()))
@@ -82,7 +84,11 @@ func (m *Meta) Bytes() ([]byte, error) {
 	t := apimetav1.Time{}
 
 	switch v := m.obj.(type) {
+	case *v1alpha1.Configuration:
+		t = v.GetCreationTimestamp()
 	case *v1.Configuration:
+		t = v.GetCreationTimestamp()
+	case *v1alpha1.Provider:
 		t = v.GetCreationTimestamp()
 	case *v1.Provider:
 		t = v.GetCreationTimestamp()
@@ -101,10 +107,13 @@ func (m *Meta) Bytes() ([]byte, error) {
 	return data, nil
 }
 
+// upsertDeps takes a v1beta1.Dependency and a runtime.Object of type that can
+// be converted to a v1.Pkg and returns an updated runtime.Object with a slice
+// of dependencies that includes the provided dependency d.
 func upsertDeps(d v1beta1.Dependency, o runtime.Object) error { // nolint:gocyclo
-	p, ok := o.(v1.Pkg)
+	p, ok := xpkg.TryConvertToPkg(o, &v1.Provider{}, &v1.Configuration{})
 	if !ok {
-		return errors.New(errNotAPackage)
+		return errors.New(errUnsupportedPackageVersion)
 	}
 	deps := p.GetDependencies()
 
@@ -142,9 +151,13 @@ func upsertDeps(d v1beta1.Dependency, o runtime.Object) error { // nolint:gocycl
 		deps = append(deps, dep)
 	}
 
-	switch v := p.(type) {
+	switch v := o.(type) {
+	case *v1alpha1.Configuration:
+		v.Spec.DependsOn = convertToV1alpha1(deps)
 	case *v1.Configuration:
 		v.Spec.DependsOn = deps
+	case *v1alpha1.Provider:
+		v.Spec.DependsOn = convertToV1alpha1(deps)
 	case *v1.Provider:
 		v.Spec.DependsOn = deps
 	}
@@ -169,4 +182,12 @@ func cleanNullTs(p runtime.Object) ([]byte, error) {
 	delete(m["metadata"].(map[string]interface{}), "creationTimestamp")
 
 	return sigsyaml.Marshal(m)
+}
+
+func convertToV1alpha1(deps []v1.Dependency) []v1alpha1.Dependency {
+	alphaDeps := make([]v1alpha1.Dependency, 0)
+	for _, d := range deps {
+		alphaDeps = append(alphaDeps, manager.ConvertToV1alpha1(d))
+	}
+	return alphaDeps
 }

--- a/internal/xpkg/workspace/meta/meta_test.go
+++ b/internal/xpkg/workspace/meta/meta_test.go
@@ -26,8 +26,10 @@ import (
 
 	"github.com/crossplane/crossplane-runtime/pkg/test"
 	metav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	metav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
 	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
 
+	"github.com/upbound/up/internal/xpkg"
 	"github.com/upbound/up/internal/xpkg/dep"
 	"github.com/upbound/up/internal/xpkg/dep/resolver/image"
 )
@@ -82,6 +84,40 @@ func TestUpsert(t *testing.T) {
 				},
 			},
 		},
+		"AddEntryNoPriorV1alpha1": {
+			reason: "Should not return an error if package is created at path.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-gcp@v1.0.0",
+					string(v1beta1.ConfigurationPackageType),
+				),
+				metaFile: &metav1alpha1.Configuration{
+					TypeMeta: apimetav1.TypeMeta{
+						APIVersion: "meta.pkg.crossplane.io/v1",
+						Kind:       "Configuration",
+					},
+					ObjectMeta: apimetav1.ObjectMeta{
+						Name: "getting-started-with-aws",
+					},
+					Spec: metav1alpha1.ConfigurationSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							Crossplane: &metav1alpha1.CrossplaneConstraints{
+								Version: ">=1.0.0-0",
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []v1beta1.Dependency{
+					{
+						Package:     "crossplane/provider-gcp",
+						Type:        v1beta1.ConfigurationPackageType,
+						Constraints: "v1.0.0",
+					},
+				},
+			},
+		},
 		"InsertNewEntry": {
 			reason: "Should not return an error if package is created at path.",
 			args: args{
@@ -103,6 +139,51 @@ func TestUpsert(t *testing.T) {
 								Version: ">=1.0.0-0",
 							},
 							DependsOn: []metav1.Dependency{
+								{
+									Provider: pointer.String("crossplane/provider-aws"),
+									Version:  ">=1.0.5",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []v1beta1.Dependency{
+					{
+						Package:     "crossplane/provider-aws",
+						Type:        v1beta1.ProviderPackageType,
+						Constraints: ">=1.0.5",
+					},
+					{
+						Package:     "crossplane/provider-gcp",
+						Type:        v1beta1.ProviderPackageType,
+						Constraints: "v1.0.0",
+					},
+				},
+			},
+		},
+		"InsertNewEntryV1alpha1": {
+			reason: "Should not return an error if package is created at path.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-gcp@v1.0.0",
+					string(v1beta1.ProviderPackageType),
+				),
+				metaFile: &metav1alpha1.Provider{
+					TypeMeta: apimetav1.TypeMeta{
+						APIVersion: "meta.pkg.crossplane.io/v1",
+						Kind:       "Provider",
+					},
+					ObjectMeta: apimetav1.ObjectMeta{
+						Name: "getting-started-with-aws",
+					},
+					Spec: metav1alpha1.ProviderSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							Crossplane: &metav1alpha1.CrossplaneConstraints{
+								Version: ">=1.0.0-0",
+							},
+							DependsOn: []metav1alpha1.Dependency{
 								{
 									Provider: pointer.String("crossplane/provider-aws"),
 									Version:  ">=1.0.5",
@@ -192,6 +273,30 @@ func TestUpsertDeps(t *testing.T) {
 				},
 			},
 		},
+		"EmptyDependencyListV1alpha1": {
+			reason: "Should return an updated deps list with the included provider, for v1alpha1.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-aws@v1.0.0",
+					string(v1beta1.ProviderPackageType),
+				),
+				pkg: &metav1alpha1.Configuration{
+					Spec: metav1alpha1.ConfigurationSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							DependsOn: []metav1alpha1.Dependency{},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []metav1.Dependency{
+					{
+						Provider: pointer.StringPtr("crossplane/provider-aws"),
+						Version:  "v1.0.0",
+					},
+				},
+			},
+		},
 		"InsertIntoDependencyList": {
 			reason: "Should return an updated deps list with 2 entries.",
 			args: args{
@@ -203,6 +308,39 @@ func TestUpsertDeps(t *testing.T) {
 					Spec: metav1.ConfigurationSpec{
 						MetaSpec: metav1.MetaSpec{
 							DependsOn: []metav1.Dependency{
+								{
+									Configuration: pointer.String("crossplane/provider-aws"),
+									Version:       "v1.0.0",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []metav1.Dependency{
+					{
+						Configuration: pointer.String("crossplane/provider-aws"),
+						Version:       "v1.0.0",
+					},
+					{
+						Provider: pointer.String("crossplane/provider-gcp"),
+						Version:  "v1.0.1",
+					},
+				},
+			},
+		},
+		"InsertIntoDependencyListV1alpha1": {
+			reason: "Should return an updated deps list with 2 entries, even for v1alpha packages.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-gcp@v1.0.1",
+					string(v1beta1.ProviderPackageType),
+				),
+				pkg: &metav1alpha1.Configuration{
+					Spec: metav1alpha1.ConfigurationSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							DependsOn: []metav1alpha1.Dependency{
 								{
 									Configuration: pointer.String("crossplane/provider-aws"),
 									Version:       "v1.0.0",
@@ -254,6 +392,35 @@ func TestUpsertDeps(t *testing.T) {
 				},
 			},
 		},
+		"UpdateDependencyListV1alpha1": {
+			reason: "Should return an updated deps list with the provider version updated, for v1alpha1.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-aws@v1.0.1",
+					string(v1beta1.ConfigurationPackageType),
+				),
+				pkg: &metav1alpha1.Provider{
+					Spec: metav1alpha1.ProviderSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							DependsOn: []metav1alpha1.Dependency{
+								{
+									Configuration: pointer.String("crossplane/provider-aws"),
+									Version:       "v1.0.0",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []metav1.Dependency{
+					{
+						Configuration: pointer.String("crossplane/provider-aws"),
+						Version:       "v1.0.1",
+					},
+				},
+			},
+		},
 		"UseDefaultTag": {
 			reason: "Should return an error indicating the package name is invalid.",
 			args: args{
@@ -265,6 +432,35 @@ func TestUpsertDeps(t *testing.T) {
 					Spec: metav1.ProviderSpec{
 						MetaSpec: metav1.MetaSpec{
 							DependsOn: []metav1.Dependency{
+								{
+									Provider: pointer.String("crossplane/provider-aws"),
+									Version:  "v1.0.0",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []metav1.Dependency{
+					{
+						Provider: pointer.String("crossplane/provider-aws"),
+						Version:  image.DefaultVer,
+					},
+				},
+			},
+		},
+		"UseDefaultTagV1alpha1": {
+			reason: "Should return an error indicating the package name is invalid, for v1alpha1.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-aws",
+					string(v1beta1.ProviderPackageType),
+				),
+				pkg: &metav1alpha1.Provider{
+					Spec: metav1alpha1.ProviderSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							DependsOn: []metav1alpha1.Dependency{
 								{
 									Provider: pointer.String("crossplane/provider-aws"),
 									Version:  "v1.0.0",
@@ -311,6 +507,34 @@ func TestUpsertDeps(t *testing.T) {
 				err: errors.New(errMetaContainsDupeDep),
 			},
 		},
+		"DuplicateDepV1alpha1": {
+			reason: "Should return an error indicating duplicate dependencies detected, v1alpha1.",
+			args: args{
+				dep: dep.NewWithType(
+					"crossplane/provider-aws",
+					string(v1beta1.ProviderPackageType),
+				),
+				pkg: &metav1alpha1.Provider{
+					Spec: metav1alpha1.ProviderSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							DependsOn: []metav1alpha1.Dependency{
+								{
+									Provider: pointer.String("crossplane/provider-aws"),
+									Version:  "v1.0.0",
+								},
+								{
+									Provider: pointer.String("crossplane/provider-aws"),
+									Version:  "v1.0.1",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				err: errors.New(errMetaContainsDupeDep),
+			},
+		},
 	}
 
 	for name, tc := range cases {
@@ -323,7 +547,7 @@ func TestUpsertDeps(t *testing.T) {
 			}
 
 			if tc.want.deps != nil {
-				p := tc.args.pkg.(metav1.Pkg)
+				p, _ := xpkg.TryConvertToPkg(tc.args.pkg, &metav1.Provider{}, &metav1.Configuration{})
 				if diff := cmp.Diff(tc.want.deps, p.GetDependencies()); diff != "" {
 					t.Errorf("\n%s\nUpsertDeps(...): -want err, +got err:\n%s", tc.reason, diff)
 				}
@@ -364,6 +588,42 @@ func TestDependsOn(t *testing.T) {
 								Version: ">=1.0.0-0",
 							},
 							DependsOn: []metav1.Dependency{
+								{
+									Provider: pointer.String("crossplane/provider-aws"),
+									Version:  "v1.0.0",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				deps: []v1beta1.Dependency{
+					{
+						Package:     "crossplane/provider-aws",
+						Type:        v1beta1.ProviderPackageType,
+						Constraints: "v1.0.0",
+					},
+				},
+			},
+		},
+		"SingleDependencyV1alpha1": {
+			reason: "Should return a slice with a single entry even for v1alpha1 files.",
+			args: args{
+				metaFile: &metav1alpha1.Configuration{
+					TypeMeta: apimetav1.TypeMeta{
+						APIVersion: "meta.pkg.crossplane.io/v1",
+						Kind:       "Configuration",
+					},
+					ObjectMeta: apimetav1.ObjectMeta{
+						Name: "getting-started-with-aws",
+					},
+					Spec: metav1alpha1.ConfigurationSpec{
+						MetaSpec: metav1alpha1.MetaSpec{
+							Crossplane: &metav1alpha1.CrossplaneConstraints{
+								Version: ">=1.0.0-0",
+							},
+							DependsOn: []metav1alpha1.Dependency{
 								{
 									Provider: pointer.String("crossplane/provider-aws"),
 									Version:  "v1.0.0",

